src/shm/shm.c

```diff
@@ -0,0 +1,470 @@
+#include "shm.h"
+#include <sys/mman.h>
+#include <sys/file.h>
+#include <errno.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+
+mmap_creat_ret_t init_or_create_mmap(char* passwd){
+    char* path = malloc(strlen("/ocp_sync")+strlen(passwd)+1);
+    strcpy(path, "/ocp_sync");
+    strcat(path, passwd);
+    int fdm = shm_open(path, O_CREAT | O_EXCL | O_RDWR | O_TRUNC, 0660);
+    printf("opn_sync %d %d\n", fdm, errno);
+    if (fdm == -1){
+        if (errno == EEXIST){
+            fdm = shm_open(path, O_RDWR, 0);
+            coord_struct* str = (coord_struct*) mmap(NULL, sizeof(coord_struct), PROT_READ | PROT_WRITE, MAP_SHARED, fdm, 0);
+            printf("mmp sync %p %d\n", str, errno);
+            return (mmap_creat_ret_t){str, fdm, path};
+        } else {
+            err_and_leave("Error when accessing shared memory", 5);
+        }
+    }
+    ftruncate(fdm, sizeof(coord_struct));
+    coord_struct* str = /*malloc(sizeof(coord_struct));/*/ (coord_struct*) mmap(NULL, sizeof(coord_struct), PROT_READ | PROT_WRITE, MAP_SHARED, fdm, 0);
+    // TODO: init all locks and the like here, returning will relinquish the lock and others might access
+    init_mutex(&(str->lock));
+    init_mutex(&(str->sender_lock));
+    init_mutex(&(str->receiver_lock));
+    init_cond_pack(&(str->reader_arrived));
+    init_cond_pack(&(str->writer_ready));
+
+    return (mmap_creat_ret_t){str, fdm, path};
+}
+
+copy_struct* init_copy(size_t width, int fd){
+    size_t size = sizeof(copy_struct) + (2 * width);
+    ftruncate(fd, size);
+    copy_struct* copy_mem = (copy_struct*) mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
+    printf("cmm %p %d\n", copy_mem, errno);
+    init_cond_pack(&(copy_mem->reader_cnt));
+    init_cond_pack(&(copy_mem->signal_rdr));
+
+    init_mutex(&(copy_mem->started_lock));
+    init_mutex(&(copy_mem->joining_lock));
+
+    init_rwlock(&(copy_mem->active[0]));
+    init_rwlock(&(copy_mem->active[1]));
+    init_rwlock(&(copy_mem->leaving[0]));
+    init_rwlock(&(copy_mem->leaving[1]));
```

Comment:
These arrays aren't of length 2 because there is 2 readers, the copy buffer is divided in two areas so that while the sender is writing to one area the readers can all read from the other one to minimize time spent waiting for the other processes to finish working

There's only ever a need for two of these areas (I'm assuming reading and writing take similar amounts of time) so even with an arbitrary number of readers only 4 inits will be needed